test/unit-test/sigv4_utest.c

```diff
@@ -214,3 +532,725 @@ void test_SigV4_AwsIotDateToIso8601_Formatting_Error()
+                   HTTP_EMPTY_PATH_LEN - LINEFEED_CHAR_LEN;
+    longQuery = malloc( longQueryLen );
+    TEST_ASSERT_NOT_NULL( longQuery );
+    /* Populate a long query parameter name. */
+    memset( longQuery, ( int ) 'P', longQueryLen );
+    resetInputParams();
+    params.pHttpParameters->pPath = NULL;
+    params.pHttpParameters->pathLen = 0U;
+    params.pHttpParameters->pQuery = longQuery;
+    params.pHttpParameters->queryLen = longQueryLen;
+    returnStatus = SigV4_GenerateHTTPAuthorization( &params, authBuf, &authBufLen, &signature, &signatureLen );
+    TEST_ASSERT_EQUAL( SigV4InsufficientMemory, returnStatus );
+    free( longQuery );
+
+    /* Test case when there is insufficient processing buffer space for writing signing key.
+     * This case is created by using a long Region value that causes the "String to Sign" data
+     * to crowd out space for the Signing Key. */
+    size_t lenOfStringToSignWithoutRegion = ( STRING_TO_SIGN_LEN_WITH_DEFAULT_REGION ) -strlen( REGION );
+    size_t longRegionLen = SIGV4_PROCESSING_BUFFER_LENGTH - lenOfStringToSignWithoutRegion;
+    /*size_t longRegionLen = SIGV4_PROCESSING_BUFFER_LENGTH - 4 * SIGV4_HASH_MAX_DIGEST_LENGTH; */
+    longRegion = malloc( longRegionLen );
+    /* Fill gibberish in the long region name. */
+    memset( longRegion, ( int ) 'R', longRegionLen );
+    resetInputParams();
+    params.pRegion = longRegion;
+    params.regionLen = longRegionLen;
+    returnStatus = SigV4_GenerateHTTPAuthorization( &params, authBuf, &authBufLen, &signature, &signatureLen );
+    TEST_ASSERT_EQUAL( SigV4InsufficientMemory, returnStatus );
+    free( longRegion );
+}
+
+/* Test that the library can encode non-alphanumeric characters in a query string. */
+void test_SigV4_GenerateHTTPAuthorization_Encode_URI_Non_AlphaNumeric()
+{
+    SigV4Status_t returnStatus;
+
+    resetInputParams();
+    params.pHttpParameters->pQuery = QUERY_WITH_NON_ALPHA_NUMBERIC_CHARS;
+    params.pHttpParameters->queryLen = strlen( QUERY_WITH_NON_ALPHA_NUMBERIC_CHARS );
+    returnStatus = SigV4_GenerateHTTPAuthorization( &params, authBuf, &authBufLen, &signature, &signatureLen );
+    TEST_ASSERT_EQUAL( SigV4Success, returnStatus );
+
+    params.pHttpParameters->pQuery = QUERY_WITH_SPECIAL_CHARS;
+    params.pHttpParameters->queryLen = strlen( QUERY_WITH_SPECIAL_CHARS );
+    returnStatus = SigV4_GenerateHTTPAuthorization( &params, authBuf, &authBufLen, &signature, &signatureLen );
+    TEST_ASSERT_EQUAL( SigV4Success, returnStatus );
+}
+
+/**
+ * @brief Coverage for cases when the hashing functions return errors.
+ */
+void test_SigV4_GenerateHTTPAuthorization_Hash_Errors()
+{
+    SigV4Status_t returnStatus;
+    size_t i;
+
+    params.pCredentials->pSecretAccessKey = SECRET_KEY_LONGER_THAN_HASH_BLOCK;
+    params.pCredentials->secretAccessKeyLen = strlen( SECRET_KEY_LONGER_THAN_HASH_BLOCK );
+
+    for( i = 0U; i < HAPPY_PATH_HASH_ITERATIONS; i++ )
```

Comment:
This is certainly a hack for now. I'd just leave a comment if you run out of time in actually being able to assert what iteration you expect to fail.